prediction/src/preprocess/improved_lung_segmentation.py

```diff
@@ -0,0 +1,444 @@
+
+    Returns:
+        np.ndarray[bool]: patient - converted Lungs
+    """
+    z_coord, y_coord, x_coord = np.where(trachea)
+    z, y, x = np.where(patient)
+
+    z = np.unique(z)
+    z_mean = np.mean(np.unique(z_coord), axis=0)
+    z_down = abs(z[0] - z_mean)
+    z_up = abs(z[-1] - z_mean)
+    if z_down > z_up:
+        patient = np.flip(patient, axis=0)
+    return patient
+
+
+def conventional_lung_segmentation(patient):
+    """ The lungs segmentation.
+
+    The algorithm consists of the following steps:
+        (1)Extraction of large airways;
+        (2)Segmentation of lung regions;
+        (3)Separation of the left and right lungs;
+        (4)Smoothing.
+
+    The function aggregates extract_seeds, growing_bronchis,
+    flip_lung, grow_lungs, lungs_postprocessing.
+
+    Args:
+        patient (np.ndarray[bool]): a label of CT
+
+    Returns:
+        np.ndarray[bool]: lungs_seeds - a label with the segmented lungs.
+        np.ndarray[bool]: trachea - a label with the segmented trachea.
+    """
+
+    seeds = extract_seeds(patient)
+    trachea, _ = growing_bronchis(patient, seeds)
+#     patient = flip_lung(patient, trachea)
+
+#     seeds = extract_seeds(patient)
+#     trachea, _ = growing_bronchis(patient, seeds)
+
+    lungs_seeds = grow_lungs(patient, trachea)
+
+    selem = skimage.morphology.ball(int(patient.shape[-1] * .01))
+    lungs_seeds = skimage.morphology.binary_closing(lungs_seeds, selem)
+
+    lungs_seeds = lungs_postprocessing(lungs_seeds)
+
+    return lungs_seeds, trachea
+
+
+def cumulation(lung):
+    """The cumulative x-position inside a lung.
+
+    Args:
+        lung: The segmented lung (left or right)
+
+    Returns:
+        lung: cumulative x-position
+        z_coords: z-coordinates in lung
+
+    """
+    z_coords, y_coords, x_coords = np.where(lung)
+    current_volume = 0
+    volume = np.count_nonzero(lung)
+    for coord in np.unique(x_coords):
+        current_volume += np.count_nonzero(lung[:, :, coord])
+        lung[:, :, coord][lung[:, :, coord] != 0] = current_volume / float(volume)
+    return lung, z_coords
+
+
+def ventricular_extraction(err_lung):
+    """ The extraction of the ventricle.
+
+    On all CT there is a ventricle. In order for the algorithm
+    doesn't define the ventricle as a segmentation error,
+    it is necessary to remove the error mask of the largest
+    size that corresponds to the ventricle.
+
+    Args:
+        err_lung (np.ndarray[bool]): the label of errors in the lung.
+
+    Returns:
+        np.ndarray[bool]: err_lung - the label of errors in the lung without the ventricle.
+
+    """
+    max_marker = -1
+    z_max = -1
+    lav, mar = scipy.ndimage.label(err_lung)
+    volumes = np.hstack([np.bincount(lav.flatten()), [-1]])
+
+    for i in range(1, mar + 1):
+        z, y, x = np.where(lav == i)
+        curr_z = z.max()
+        if curr_z > z_max:
+            curr_z = z_max
+            coords = (z, y, x)
+
+        if curr_z == z_max:
```

Comment:
How about `if (curr_z == z_max) and (volumes[i] > volumes[max_marker]):`?

Reply:
Thank you :)